File: software/ai/sdk-go/x402/payment.go
package x402

import (
	"crypto/ecdsa"
	"encoding/json"
	"errors"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/math"
	"github.com/ethereum/go-ethereum/crypto"
)

func SignEIP3009Authorization(
	privateKey *ecdsa.PrivateKey,
	tokenAddress common.Address,
	from common.Address,
	to common.Address,
	value *big.Int,
	validAfter *big.Int,
	validBefore *big.Int,
	nonce [32]byte,
	chainID *big.Int,
) ([]byte, error) {
	domainSeparator := buildDomainSeparator(tokenAddress, chainID)

	transferTypeHash := crypto.Keccak256([]byte(
		"TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)",
	))

	structHash := crypto.Keccak256(
		transferTypeHash,
		common.LeftPadBytes(from.Bytes(), 32),
		common.LeftPadBytes(to.Bytes(), 32),
		math.U256Bytes(value),
		math.U256Bytes(validAfter),
		math.U256Bytes(validBefore),
		nonce[:],
	)

	digestInput := make([]byte, 0, 66)
	digestInput = append(digestInput, 0x19, 0x01)
	digestInput = append(digestInput, domainSeparator...)
	digestInput = append(digestInput, structHash...)
	digest := crypto.Keccak256(digestInput)

	sig, err := crypto.Sign(digest, privateKey)
	if err != nil {
		return nil, err
	}

	if sig[64] < 27 {
		sig[64] += 27
	}

	return sig, nil
}

func buildDomainSeparator(tokenAddress common.Address, chainID *big.Int) []byte {
	domainTypeHash := crypto.Keccak256([]byte(
		"EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)",
	))

	nameHash := crypto.Keccak256([]byte("USD Coin"))
	versionHash := crypto.Keccak256([]byte("2"))

	return crypto.Keccak256(
		domainTypeHash,
		nameHash,
		versionHash,
		math.U256Bytes(chainID),
		common.LeftPadBytes(tokenAddress.Bytes(), 32),
	)
}

func BuildPaymentHeader(
	privateKey *ecdsa.PrivateKey,
	req *PaymentRequirement,
	from common.Address,
	chainID *big.Int,
) (string, error) {
	if req == nil {
		return "", errors.New("nil payment requirement")
	}

	amount, ok := new(big.Int).SetString(req.MaxAmountRequired, 10)
	if !ok {
		return "", errors.New("invalid amount")
	}

	deadline, ok := new(big.Int).SetString(req.RequiredDeadline, 10)
	if !ok {
		deadline = big.NewInt(0)
	}

	var nonce [32]byte
	nonceHash := crypto.Keccak256(
		from.Bytes(),
		req.PayTo.Bytes(),
		math.U256Bytes(amount),
	)
	copy(nonce[:], nonceHash[:32])

	tokenAddress := common.HexToAddress("0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913")

	sig, err := SignEIP3009Authorization(
		privateKey,
		tokenAddress,
		from,
		req.PayTo,
		amount,
		big.NewInt(0),
		deadline,
		nonce,
		chainID,
	)
	if err != nil {
		return "", err
	}

	payload := map[string]interface{}{
		"x402Version": 1,
		"scheme":      req.Scheme,
		"network":     req.Network,
		"payload": map[string]interface{}{
			"signature":   common.Bytes2Hex(sig),
			"from":        from.Hex(),
			"to":          req.PayTo.Hex(),
			"value":       req.MaxAmountRequired,
			"validAfter":  "0",
			"validBefore": req.RequiredDeadline,
			"nonce":       common.Bytes2Hex(nonce[:]),
		},
	}

	headerBytes, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	return string(headerBytes), nil
}

func ParsePaymentRequirements(data []byte) ([]PaymentRequirement, error) {
	var requirements []PaymentRequirement
	if err := json.Unmarshal(data, &requirements); err != nil {
		var single PaymentRequirement
		if err2 := json.Unmarshal(data, &single); err2 != nil {
			return nil, errors.New("invalid payment requirements format")
		}
		return []PaymentRequirement{single}, nil
	}
	return requirements, nil
}

File: software/ai/sdk-ts/src/x402/payment.ts
import {
  type Address,
  type Hex,
  type LocalAccount,
  encodePacked,
  keccak256,
  toHex,
  encodeAbiParameters,
  parseAbiParameters,
} from "viem";
import type { EIP3009Authorization, PaymentRequirement } from "../types/x402.js";
import { generateNonce } from "../utils/encoding.js";

const EIP3009_TYPEHASH = keccak256(
  encodePacked(
    ["string"],
    ["TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)"]
  )
);

export async function signEIP3009Authorization(
  account: LocalAccount,
  params: {
    tokenAddress: Address;
    from: Address;
    to: Address;
    value: bigint;
    validAfter: bigint;
    validBefore: bigint;
    chainId: number;
  }
): Promise<{ authorization: EIP3009Authorization; signature: Hex }> {
  const nonce = generateNonce();

  const authorization: EIP3009Authorization = {
    from: params.from,
    to: params.to,
    value: params.value,
    validAfter: params.validAfter,
    validBefore: params.validBefore,
    nonce,
  };

  const domain = {
    name: "USD Coin",
    version: "2",
    chainId: params.chainId,
    verifyingContract: params.tokenAddress,
  } as const;

  const types = {
    TransferWithAuthorization: [
      { name: "from", type: "address" },
      { name: "to", type: "address" },
      { name: "value", type: "uint256" },
      { name: "validAfter", type: "uint256" },
      { name: "validBefore", type: "uint256" },
      { name: "nonce", type: "bytes32" },
    ],
  } as const;

  const signature = await account.signTypedData({
    domain,
    types,
    primaryType: "TransferWithAuthorization",
    message: {
      from: authorization.from,
      to: authorization.to,
      value: authorization.value,
      validAfter: authorization.validAfter,
      validBefore: authorization.validBefore,
      nonce: authorization.nonce,
    },
  });

  return { authorization, signature };
}

export function buildPaymentHeader(
  authorization: EIP3009Authorization,
  signature: Hex,
  requirement: PaymentRequirement
): string {
  const payload = {
    x402Version: 1,
    scheme: requirement.scheme,
    network: requirement.network,
    payload: {
      signature,
      authorization: {
        from: authorization.from,
        to: authorization.to,
        value: authorization.value.toString(),
        validAfter: authorization.validAfter.toString(),
        validBefore: authorization.validBefore.toString(),
        nonce: authorization.nonce,
      },
    },
  };

  return btoa(JSON.stringify(payload));
}

export function parsePaymentHeader(header: string): {
  authorization: EIP3009Authorization;
  signature: Hex;
  scheme: string;
  network: string;
} {
  const decoded = JSON.parse(atob(header));

  return {
    authorization: {
      from: decoded.payload.authorization.from as Address,
      to: decoded.payload.authorization.to as Address,
      value: BigInt(decoded.payload.authorization.value),
      validAfter: BigInt(decoded.payload.authorization.validAfter),
      validBefore: BigInt(decoded.payload.authorization.validBefore),
      nonce: decoded.payload.authorization.nonce as Hex,
    },
    signature: decoded.payload.signature as Hex,
    scheme: decoded.scheme,
    network: decoded.network,
  };
}
